align/pnr/hpwl.py

```diff
@@ -94,13 +95,55 @@ def aux( module, prefix_path, translate_d):
     return nets_d
 
 
-def calculate_HPWL_from_placement_verilog_d( placement_verilog_d, concrete_name):
+def calculate_HPWL_from_placement_verilog_d( placement_verilog_d, concrete_name, nets_d):
 
     """Two ways to do this. Compute hierarchically or virtually flat."""
 
     """Find all the leaf terminals"""
 
     nets_d = gen_netlist( placement_verilog_d, concrete_name)
+    modules = { module['concrete_name']: module for module in placement_verilog_d['modules']}
+    instances = { (module['concrete_name'],instance['instance_name']): instance for module in placement_verilog_d['modules'] for instance in module['instances']}
+
+    leaf_terminals = defaultdict(list)
+
+    for leaf in placement_verilog_d['leaves']:
+        ctn = leaf['concrete_name']
+        for terminal in leaf['terminals']:
+            leaf_terminals[(ctn,terminal['name'])].append( terminal['rect'])
+
+    logger_fn = logger.debug
 
     HPWL = 0
+
+    for hnet, hpins in nets_d.items():
+
+        mx, Mx, my, My = None, None, None, None
+
+        for hpin in hpins:
+            ctn = concrete_name
+            tr = Transformation()
+            for instance_name in hpin[:-1]:
+                instance = instances[(ctn,instance_name)]
+                ctn = instance['concrete_template_name']
+                tr = tr.postMult(Transformation( **instance['transformation']))
+            
+            for r in leaf_terminals[(ctn,hpin[-1])]:
+                [x0, y0, x1, y1] = tr.hitRect( Rect(*r)).canonical().toList()
+
+                #[x0, y0, x1, y1] = [ (x0+x1)//2, (y0+y1)//2, (x0+x1)//2, (y0+y1)//2]
```

Comment:
Uncommenting this line and filtering the power nets matches with the placer HPWL.

Reply:
No, I don't think so. That was my attempt to make them match, but they still don't match.
There are several reason for this as far as I can tell:
- I've defined HPWL as the rectangle around all the terminals connected to the net, which can span multiple levels of hierarchy. The internal HPWL calculator works one level at a time.
- I have designs without power nets at all that don't match.
- I don't think the center value is computed correctly in the internal HPWL calculator. The X coord is zero too often in my testcases. I'm working on try to understand this.

If you want to look at an example, you can try:
```
pytest -- tests/pnr/test_entrypoint2.py 
(cd $ALIGN_WORK_DIR/row_entrypoint2/; schematic2layout.py inputs/ -s matrix --flow_start 3_pnr -n 40 --gui )
```
The individual wirelengths are simple enough to compute by hand for these examples.